pkg/util/containers/v2/metrics/collector_containerd.go

```
@@ -0,0 +1,416 @@
+		cpuLimitPct = float64(*cpuLimits.Quota) / float64(*cpuLimits.Period)
+	}
+
+	limit := cpuLimitPct * timeDiff
+	return &limit
+}
+
+func getContainerdStatsWindows(windowsStats *wstats.WindowsContainerStatistics) *ContainerStats {
+	if windowsStats == nil {
+		return nil
+	}
+
+	return &ContainerStats{
+		Timestamp: windowsStats.Timestamp,
+		CPU:       getContainerdCPUStatsWindows(windowsStats.Processor),
+		Memory:    getContainerdMemoryStatsWindows(windowsStats.Memory),
+		IO:        getContainerdIOStatsWindows(windowsStats.Storage),
+	}
+}
+
+func getContainerdCPUStatsWindows(procStats *wstats.WindowsContainerProcessorStatistics) *ContainerCPUStats {
+	if procStats == nil {
+		return nil
+	}
+
+	return &ContainerCPUStats{
+		Total:  uint64ToFloat64Ptr(procStats.TotalRuntimeNS),
+		System: uint64ToFloat64Ptr(procStats.RuntimeKernelNS),
+		User:   uint64ToFloat64Ptr(procStats.RuntimeUserNS),
+	}
+}
+
+func getContainerdMemoryStatsWindows(memStats *wstats.WindowsContainerMemoryStatistics) *ContainerMemStats {
+	if memStats == nil {
+		return nil
+	}
+
+	return &ContainerMemStats{
+		PrivateWorkingSet: uint64ToFloat64Ptr(memStats.MemoryUsagePrivateWorkingSetBytes),
+		CommitBytes:       uint64ToFloat64Ptr(memStats.MemoryUsageCommitBytes),
+		CommitPeakBytes:   uint64ToFloat64Ptr(memStats.MemoryUsageCommitPeakBytes),
+	}
+}
+
+func getContainerdIOStatsWindows(ioStats *wstats.WindowsContainerStorageStatistics) *ContainerIOStats {
+	if ioStats == nil {
+		return nil
+	}
+
+	return &ContainerIOStats{
+		ReadBytes:       uint64ToFloat64Ptr(ioStats.ReadSizeBytes),
+		WriteBytes:      uint64ToFloat64Ptr(ioStats.WriteSizeBytes),
+		ReadOperations:  uint64ToFloat64Ptr(ioStats.ReadCountNormalized),
+		WriteOperations: uint64ToFloat64Ptr(ioStats.WriteCountNormalized),
+	}
+}
+
+func getContainerdNetworkStatsWindows() *ContainerNetworkStats {
+	// Network stats are not exposed on Windows, so return 0's for all the
+	// fields.
```

why not nil? I'd rather not report these metrics if they're not available

---

I wasn't sure about what to return in this case. I think it depends on how the caller of the collectors works and I'm not too familiar with that part of the code yet.

Should I just return something like: `nil, fmt.Error('not available on Windows)` in the interface function `GetContainerNetworkStats` ? Or `nil, nil` and log (debug lvl) the "not available on Windows" message?

---

Currently the part of the check that should call `GetContainerNetworkStats` is not implemented yet. Ideally `GetContainerNetworkStats` should return a typed error `ErrNotSupported` in this case. cc @vboulineau for visibility.

---

I think that having an `ErrNotSupported` defined in the `containers/v2/metrics` package would be great :+1:

---

It should just report nil, it's not supported and we'll never get metric, I don't think it's required to return an error there.

---

It's just that the consumer can print a meaningful log with the right level based on the returned error type. Returning `nil`, `nil` can easily backfire 😅

---

Open to both honestly.
I understand the `nil, nil` issue, which is not always expected as an API.
On the other hand, I think it makes sense for this particular API, we used pointers to highlight that any data can be missing, and it makes sense to not return any data.
I think checking for `nil` if more straightforward than checking everywhere `if err == ErrNotSupported || err == nil {} else {}` or similar.